infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/WorkspaceServiceAccount.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.openshift.project;
+
+import io.fabric8.kubernetes.api.model.ObjectReferenceBuilder;
+import io.fabric8.openshift.api.model.PolicyRuleBuilder;
+import io.fabric8.openshift.api.model.Role;
+import io.fabric8.openshift.api.model.RoleBuilder;
+import io.fabric8.openshift.client.OpenShiftClient;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.openshift.OpenShiftClientFactory;
+
+/**
+ * Holds logic for preparing workspace service account.
+ *
+ * <p>It checks that required service account, roles and role bindings exist and creates if needed.
+ *
+ * @author Sergii Leshchenko
+ */
+class WorkspaceServiceAccount {
+
+  private final String projectName;
+  private final String name;
+  private final OpenShiftClientFactory clientFactory;
+  private final String workspaceId;
+
+  WorkspaceServiceAccount(
+      String workspaceId,
+      String projectName,
+      String serviceAccountName,
+      OpenShiftClientFactory clientFactory) {
+    this.workspaceId = workspaceId;
+    this.projectName = projectName;
+    this.name = serviceAccountName;
+    this.clientFactory = clientFactory;
+  }
+
+  /**
+   * Make sure that workspace service account exists and has `view` and `exec` role bindings.
+   *
+   * <p>Note that `view` role is used from cluster scope and `exec` role is created in the current
+   * namespace if does not exit.
+   *
+   * @throws InfrastructureException when any exception occurred
+   */
+  void prepare() throws InfrastructureException {
+    OpenShiftClient osClient = clientFactory.createOC(workspaceId);
+
+    if (osClient.serviceAccounts().inNamespace(projectName).withName(name).get() == null) {
+      createWorkspaceServiceAccount(osClient);
+    }
+
+    String execRoleName = "exec";
+    if (osClient.roles().inNamespace(projectName).withName(execRoleName).get() == null) {
+      createExecRole(osClient, execRoleName);
+    }
+
+    String execRoleBindingName = name + "-exec";
+    if (osClient.roleBindings().inNamespace(projectName).withName(execRoleBindingName).get()
+        == null) {
+      createExecRoleBinding(osClient, execRoleBindingName);
+    }
+
+    String viewRoleBindingName = name + "-view";
+    if (osClient.roleBindings().inNamespace(projectName).withName(viewRoleBindingName).get()
+        == null) {
+      createViewRoleBinding(osClient, viewRoleBindingName);
+    }
+  }
+
+  private void createWorkspaceServiceAccount(OpenShiftClient osClient) {
```

Comment:
Just to double check: does creating a SA / rolebindings require different privileges from what the standard Che SA might be given currently (esp. on public clusters)?

Reply:
If workspaces are configured to be created in the same namespace then Che SA needs admin rights in the current namespace (for OpenShift, now we create SA with `edit` rights in its project).
If workspaces are configured to be created in different namespaces then Che SA needs `cluster-admin` right or OAuth provider should be used, then a user should be able to create Projects and objects there.

So, I would say that `service-account` should be used only if:
1. Workspaces are created in the same namespace and service account is created during deployment.
2. Oauth provider is configured.
Otherwise, extra privileges should be provided for Che SA. Hope it will be helpful.